File: output_Tabnine/javascript/Hard/1925-find-the-shortest-superstring.js
/**https://leetcode.com/problems/find-the-shortest-superstring/ */
//Given an array of strings words, return the smallest string that contains each string in words as a substring. If there are multiple valid strings of the smallest length, return any of them.
//You may assume that no string in words is a substring of another string in words.
// 
//Example 1:
//Input: words = ["alex","loves","leetcode"]
//Output: "alexlovesleetcode"
//Explanation: All permutations of "alex","loves","leetcode" would also be accepted.
//Example 2:
//Input: words = ["catg","ctaagt","gcta","ttca","atgcatc"]
//Output: "gctaagttcatgcatc"
// 
//Constraints:
//	1 <= words.length <= 12
//	1 <= words[i].length <= 20
//	words[i] consists of lowercase English letters.
//	All the strings of words are unique.
/**
 * @param {string[]} words
 * @return {string}
 */
var shortestSuperstring = function(words) {
    let n = words.length;
    let dp = Array(1 << n).fill(Infinity);
    let parent = Array(1 << n).fill(-1);
    let prefixSum = Array(n).fill(0);
    
    // Calculate prefix sum for each word
    for (let i = 0; i < n; i++) {
        for (let j = i + 1; j < n; j++) {
            prefixSum[i] ^= prefixSum[j] ^= prefixSum[i] ^= words[j].indexOf(words[i]);
        }
    }
    
    // Build the DP table
    for (let mask = 1; mask < (1 << n); mask++) {
        for (let i = 0; i < n; i++) {
            if ((mask >> i) & 1) {
                let prevMask = mask ^ (1 << i);
                let prevCost = dp[prevMask];
                if (prevCost < Infinity) {
                    let newCost = prevCost + prefixSum[i];
                    if (newCost < dp[mask]) {
                        dp[mask] = newCost;
                        parent[mask] = prevMask;
                    }
                }
            }
        }
    }
    
    // Reconstruct the shortest superstring
    let result = '';
    let currentMask = (1 << n) - 1;
    while (currentMask > 0) {
        result = words[0] + result;
        let prevMask = parent[currentMask];
        for (let i = 0; i < n; i++) {
            if ((currentMask >> i) & 1 && (prevMask >> i) & 1) {
                result = words[i] + result;
                break;
            }
        }
        currentMask = prevMask;
    }
    
    return result;
    

};

File: output_Tabnine/python/Hard/1925-find-the-shortest-superstring.py
#https://leetcode.com/problems/find-the-shortest-superstring/
#Given an array of strings words, return the smallest string that contains each string in words as a substring. If there are multiple valid strings of the smallest length, return any of them.
#You may assume that no string in words is a substring of another string in words.
# 
#Example 1:
#Input: words = ["alex","loves","leetcode"]
#Output: "alexlovesleetcode"
#Explanation: All permutations of "alex","loves","leetcode" would also be accepted.
#Example 2:
#Input: words = ["catg","ctaagt","gcta","ttca","atgcatc"]
#Output: "gctaagttcatgcatc"
# 
#Constraints:
#	1 <= words.length <= 12
#	1 <= words[i].length <= 20
#	words[i] consists of lowercase English letters.
#	All the strings of words are unique.
class Solution(object):
    def shortestSuperstring(self, words):
        """
        :type words: List[str]
        :rtype: str
        """
        # Build a graph where each node represents a word and each edge represents a common substring
        graph = {}
        for i in range(len(words)):
            for j in range(i+1, len(words)):
                common = self.find_common_suffix(words[i], words[j])
                if common:
                    if words[i] not in graph:
                        graph[words[i]] = []
                    if words[j] not in graph:
                        graph[words[j]] = []
                    graph[words[i]].append(words[j])
                    graph[words[j]].append(words[i])
        
        # Perform a topological sort on the graph to get the order in which words can be concatenated
        order = self.topological_sort(graph)
        
        # Build the shortest superstring by concatenating the words in the order
        superstring = ""
        for word in order:
            superstring += word
            # Remove the concatenated word from the graph
            for neighbor in graph[word]:
                graph[neighbor].remove(word)
                if len(graph[neighbor]) == 0:
                    del graph[neighbor]
        
        # Concatenate the remaining words in the order
        for word in order:
            superstring += word
        
        return superstring
        